sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -864,6 +870,361 @@ def is_fast_reboot_enabled():
+                                                  self.on_port_update_event)
+
+            for lport in list(self.port_dict.keys()):
+                if self.task_stopping_event.is_set():
+                    break
+                data = self.port_dict[lport]
+                pport = int(data.get('index', '-1'))
+                channel = int(data.get('channel', '0'))
+                xcvr_type = data.get('type', None)
+                xcvr_inserted = False
+                host_tx_ready_changed = False
+                if pport < 0 or channel < 0:
+                    continue
+
+                if xcvr_type is None:
+                    # TRANSCEIVER_INFO table's 'type' is not ready, meaning xcvr is not present
+                    continue
+
+                # Procced only for 100G/40G
+                if not (xcvr_type.startswith('QSFP28') or xcvr_type.startswith('QSFP+')):
+                    continue
+
+                # Handle the case when Xcvrd was NOT running when 'host_tx_ready'
+                # was updated or this is the first run so reconcile
+                if 'host_tx_ready' not in data:
+                    data['host_tx_ready'] = self.get_host_tx_status(lport)
+
+                # It's a xcvr insertion case if TRANSCEIVER_INFO 'type' doesn't exist
+                # in previous port_dict sanpshot
+                if lport not in self.port_dict_prev or 'type' not in self.port_dict_prev[lport]:
+                    xcvr_inserted = True
+                # Check if there's an diff between current and previous host_tx_ready
+                if (lport not in self.port_dict_prev or
+                        'host_tx_ready' not in self.port_dict_prev[lport] or
+                        self.port_dict_prev[lport]['host_tx_ready'] != data['host_tx_ready']):
+                    host_tx_ready_changed = True
+                # Only proceed on xcvr insertion case or the case of host_tx_ready getting changed
+                if (not xcvr_inserted) and (not host_tx_ready_changed):
+                    continue
+                self.log_notice("{}: xcvr_inserted={}, host_tx_ready_changed={}".format(
+                    lport, xcvr_inserted, host_tx_ready_changed))
+
+                # double-check the HW presence before moving forward
+                sfp = platform_chassis.get_sfp(pport)
+                if not sfp.get_presence():
+                    self.log_error("{}: module not present!".format(lport))
+                    del self.port_dict[lport]
+                    continue
+                try:
+                    # Skip if XcvrApi is not supported
+                    api = sfp.get_xcvr_api()
+                    if api is None:
+                        self.log_error(
+                            "{}: skipping sff_mgr since no xcvr api!".format(lport))
+                        continue
+
+                    # Skip if it's not a paged memory device
+                    if api.is_flat_memory():
```

Comment:
check if it's duplicated with copper check

Reply:
is_flat API and is_copper API  are checking different fields of eeprom, so it's necessary to have both checks.
is_flat API
self.xcvr_eeprom.read(consts.FLAT_MEM_FIELD)
is_copper API                                 
self.xcvr_eeprom.read(consts.EXT_SPEC_COMPLIANCE_FIELD)